{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Level-4 data of L1000 gene-expression profiling is calculated as:\n",
    "\n",
    "Level 4 - Differential Expression (ZSPC):\n",
    "\n",
    "To obtain a measure of relative gene expression, we use a robust z-scoring procedure to generate differential expression values from normalized profiles (Level-3). We compute the **differential expression of gene x in the ith sample(xi) on the plate (zi)** as\n",
    "                              \n",
    "                              zi = xi - median(X) / 1.4826 * MAD(X)\n",
    "\n",
    "\n",
    "***where X is the vector of normalized gene expression of gene x across all samples on the plate, MAD is the median absolute deviation of X, and the factor of 1.4826 makes the denominator a consistent estimator of scale for normally distributed data.***\n",
    "\n",
    "[Subramanian et al 2017](https://www.cell.com/action/showPdf?pii=S0092-8674%2817%2931309-0)\n",
    "\n",
    "\n",
    "\n",
    "- Level-4 data - are replicate level data i.e. where you have multiple profiles been perturbed by the same compound (pertubagen)\n",
    "\n",
    "#### The goal here:\n",
    "\n",
    "\n",
    "-- is to determine the median score of each compound per dose based on taking the median of the correlation values between replicates of the same compound."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pathlib\n",
    "import requests\n",
    "import pickle\n",
    "import argparse\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import re\n",
    "from os import walk\n",
    "from collections import Counter\n",
    "import random\n",
    "import shutil\n",
    "from statistics import median\n",
    "import cmapPy.pandasGEXpress.parse_gct as pg\n",
    "from cmapPy.pandasGEXpress.parse import parse\n",
    "from io import BytesIO\n",
    "from urllib.request import urlopen\n",
    "from zipfile import ZipFile\n",
    "\n",
    "import warnings\n",
    "warnings.simplefilter(action='ignore', category=FutureWarning)\n",
    "np.warnings.filterwarnings('ignore', category=np.VisibleDeprecationWarning)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['level_4_zspc_n27837x978.gctx',\n",
       " 'level_4W_zspc_n27837x978.gctx',\n",
       " 'level_3_q2norm_n27837x978.gctx',\n",
       " 'level_5_rank_n9482x978.gctx',\n",
       " 'level_5_modz_n9482x978.gctx',\n",
       " 'col_meta_level_5_n169494.txt',\n",
       " 'level_5_modz_cid_n9482.grp',\n",
       " 'set_size_3_level_3_REP.A_A549_only_all_compounds_and_random_well_DMSO.gmt',\n",
       " 'col_meta_level_3_REP.A_A549_only_n27837.txt',\n",
       " 'REP.A_A549_pert_info.txt',\n",
       " 'level_5_modz_common_sigs_n8370x978.gctx',\n",
       " 'col_meta_level_3_n421176.txt',\n",
       " 'col_meta_level_5_REP.A_A549_only_n9482.txt']"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_dir = pathlib.Path(\"../L1000/L1000_figshare_data\")\n",
    "os.listdir(data_dir) ##files in L1000 downloaded dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "pertinfo_file = '../aligned_moa_CP_L1000.csv'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Two level-4 L1000 Data:\n",
    "\n",
    "- 'level_4W_zspc_n27837x978.gctx'\n",
    "- 'level_4_zspc_n27837x978.gctx',"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def construct_lvl4_data(data_dir, level4_dir, pertinfo_file):\n",
    "    \"\"\"\n",
    "    This function returns L1000 Level-4 Data that is aligned with \n",
    "    the important metadata information to compute compound's median scores\n",
    "    \"\"\"\n",
    "    \n",
    "    lvl4_data = parse(os.path.join(data_dir, level4_dir))\n",
    "    lvl4_data = lvl4_data.data_df.rename_axis(None).T\n",
    "    lvl4_data = lvl4_data.rename_axis(None).reset_index()\n",
    "    lvl4_data.rename(columns={\"index\": \"replicate_id\"}, inplace = True)\n",
    "    df_metalvl_5 = pd.read_csv(os.path.join(data_dir, 'col_meta_level_5_REP.A_A549_only_n9482.txt'), delimiter = \"\\t\")\n",
    "    lvl4_data['sig_id'] = lvl4_data['replicate_id'].apply(lambda x: ''.join(re.sub('(24H.+(\\_|\\.)[A-Z0-9]+.)\\:', '24H:', x)))\n",
    "    df_meta_features = df_metalvl_5[['sig_id', 'pert_id', 'pert_idose', 'det_plate', 'det_well']].copy()\n",
    "    df_meta_features['dose'] = df_meta_features['pert_idose'].map({'-666' : 0, '0.04 uM' : 1, '0.12 uM' : 2, '0.37 uM' : 3,\n",
    "                                                                   '1.11 uM' : 4, '3.33 uM' : 5, '10 uM' : 6, '20 uM' : 7})\n",
    "    df_pertinfo = pd.read_csv(pertinfo_file)\n",
    "    df_pertinfo.rename(columns={\"broad_id\": \"pert_id\"}, inplace = True)\n",
    "    df_meta_features = pd.merge(df_meta_features, df_pertinfo, on=['pert_id'], how = 'left')\n",
    "    lvl4_data = pd.merge(lvl4_data, df_meta_features, on='sig_id')\n",
    "    \n",
    "    return lvl4_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_level4 = construct_lvl4_data(data_dir, 'level_4_zspc_n27837x978.gctx', pertinfo_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>replicate_id</th>\n",
       "      <th>200814_at</th>\n",
       "      <th>222103_at</th>\n",
       "      <th>201453_x_at</th>\n",
       "      <th>204131_s_at</th>\n",
       "      <th>200059_s_at</th>\n",
       "      <th>205067_at</th>\n",
       "      <th>213702_x_at</th>\n",
       "      <th>214435_x_at</th>\n",
       "      <th>201334_s_at</th>\n",
       "      <th>...</th>\n",
       "      <th>205379_at</th>\n",
       "      <th>sig_id</th>\n",
       "      <th>pert_id</th>\n",
       "      <th>pert_idose</th>\n",
       "      <th>det_plate</th>\n",
       "      <th>det_well</th>\n",
       "      <th>dose</th>\n",
       "      <th>Metadata_broad_sample</th>\n",
       "      <th>pert_iname</th>\n",
       "      <th>moa</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>REP.A001_A549_24H_X1_B27:A03</td>\n",
       "      <td>0.3547</td>\n",
       "      <td>-0.4940</td>\n",
       "      <td>-0.1721</td>\n",
       "      <td>-0.0339</td>\n",
       "      <td>-0.4355</td>\n",
       "      <td>1.8263</td>\n",
       "      <td>-0.1316</td>\n",
       "      <td>0.0853</td>\n",
       "      <td>-0.4660</td>\n",
       "      <td>...</td>\n",
       "      <td>0.1046</td>\n",
       "      <td>REP.A001_A549_24H:A03</td>\n",
       "      <td>DMSO</td>\n",
       "      <td>-666</td>\n",
       "      <td>REP.A001_A549_24H_X1_B27|REP.A001_A549_24H_X2_...</td>\n",
       "      <td>A03</td>\n",
       "      <td>0</td>\n",
       "      <td>DMSO</td>\n",
       "      <td>DMSO</td>\n",
       "      <td>Control vehicle</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>REP.A001_A549_24H_X2_B27:A03</td>\n",
       "      <td>-0.0447</td>\n",
       "      <td>-1.6390</td>\n",
       "      <td>-0.5276</td>\n",
       "      <td>-0.5092</td>\n",
       "      <td>-0.5733</td>\n",
       "      <td>0.2445</td>\n",
       "      <td>0.6159</td>\n",
       "      <td>-1.1273</td>\n",
       "      <td>2.8250</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.5836</td>\n",
       "      <td>REP.A001_A549_24H:A03</td>\n",
       "      <td>DMSO</td>\n",
       "      <td>-666</td>\n",
       "      <td>REP.A001_A549_24H_X1_B27|REP.A001_A549_24H_X2_...</td>\n",
       "      <td>A03</td>\n",
       "      <td>0</td>\n",
       "      <td>DMSO</td>\n",
       "      <td>DMSO</td>\n",
       "      <td>Control vehicle</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>REP.A001_A549_24H_X3_B27:A03</td>\n",
       "      <td>-0.9583</td>\n",
       "      <td>0.2657</td>\n",
       "      <td>-0.8204</td>\n",
       "      <td>-0.5851</td>\n",
       "      <td>-2.0808</td>\n",
       "      <td>0.0739</td>\n",
       "      <td>-0.0497</td>\n",
       "      <td>-0.3012</td>\n",
       "      <td>-0.9787</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.1416</td>\n",
       "      <td>REP.A001_A549_24H:A03</td>\n",
       "      <td>DMSO</td>\n",
       "      <td>-666</td>\n",
       "      <td>REP.A001_A549_24H_X1_B27|REP.A001_A549_24H_X2_...</td>\n",
       "      <td>A03</td>\n",
       "      <td>0</td>\n",
       "      <td>DMSO</td>\n",
       "      <td>DMSO</td>\n",
       "      <td>Control vehicle</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>REP.A001_A549_24H_X1_B27:A04</td>\n",
       "      <td>-0.2130</td>\n",
       "      <td>0.4931</td>\n",
       "      <td>-0.8768</td>\n",
       "      <td>-0.6968</td>\n",
       "      <td>-1.7018</td>\n",
       "      <td>-0.3779</td>\n",
       "      <td>-0.6745</td>\n",
       "      <td>-1.9799</td>\n",
       "      <td>-1.1429</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.8243</td>\n",
       "      <td>REP.A001_A549_24H:A04</td>\n",
       "      <td>DMSO</td>\n",
       "      <td>-666</td>\n",
       "      <td>REP.A001_A549_24H_X1_B27|REP.A001_A549_24H_X2_...</td>\n",
       "      <td>A04</td>\n",
       "      <td>0</td>\n",
       "      <td>DMSO</td>\n",
       "      <td>DMSO</td>\n",
       "      <td>Control vehicle</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>REP.A001_A549_24H_X2_B27:A04</td>\n",
       "      <td>0.7499</td>\n",
       "      <td>-1.2819</td>\n",
       "      <td>0.4981</td>\n",
       "      <td>1.7090</td>\n",
       "      <td>1.6765</td>\n",
       "      <td>-1.2690</td>\n",
       "      <td>1.7974</td>\n",
       "      <td>-1.2213</td>\n",
       "      <td>3.4625</td>\n",
       "      <td>...</td>\n",
       "      <td>0.3157</td>\n",
       "      <td>REP.A001_A549_24H:A04</td>\n",
       "      <td>DMSO</td>\n",
       "      <td>-666</td>\n",
       "      <td>REP.A001_A549_24H_X1_B27|REP.A001_A549_24H_X2_...</td>\n",
       "      <td>A04</td>\n",
       "      <td>0</td>\n",
       "      <td>DMSO</td>\n",
       "      <td>DMSO</td>\n",
       "      <td>Control vehicle</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 988 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   replicate_id  200814_at  222103_at  201453_x_at  \\\n",
       "0  REP.A001_A549_24H_X1_B27:A03     0.3547    -0.4940      -0.1721   \n",
       "1  REP.A001_A549_24H_X2_B27:A03    -0.0447    -1.6390      -0.5276   \n",
       "2  REP.A001_A549_24H_X3_B27:A03    -0.9583     0.2657      -0.8204   \n",
       "3  REP.A001_A549_24H_X1_B27:A04    -0.2130     0.4931      -0.8768   \n",
       "4  REP.A001_A549_24H_X2_B27:A04     0.7499    -1.2819       0.4981   \n",
       "\n",
       "   204131_s_at  200059_s_at  205067_at  213702_x_at  214435_x_at  201334_s_at  \\\n",
       "0      -0.0339      -0.4355     1.8263      -0.1316       0.0853      -0.4660   \n",
       "1      -0.5092      -0.5733     0.2445       0.6159      -1.1273       2.8250   \n",
       "2      -0.5851      -2.0808     0.0739      -0.0497      -0.3012      -0.9787   \n",
       "3      -0.6968      -1.7018    -0.3779      -0.6745      -1.9799      -1.1429   \n",
       "4       1.7090       1.6765    -1.2690       1.7974      -1.2213       3.4625   \n",
       "\n",
       "   ...  205379_at                 sig_id  pert_id  pert_idose  \\\n",
       "0  ...     0.1046  REP.A001_A549_24H:A03     DMSO        -666   \n",
       "1  ...    -1.5836  REP.A001_A549_24H:A03     DMSO        -666   \n",
       "2  ...    -0.1416  REP.A001_A549_24H:A03     DMSO        -666   \n",
       "3  ...    -1.8243  REP.A001_A549_24H:A04     DMSO        -666   \n",
       "4  ...     0.3157  REP.A001_A549_24H:A04     DMSO        -666   \n",
       "\n",
       "                                           det_plate  det_well  dose  \\\n",
       "0  REP.A001_A549_24H_X1_B27|REP.A001_A549_24H_X2_...       A03     0   \n",
       "1  REP.A001_A549_24H_X1_B27|REP.A001_A549_24H_X2_...       A03     0   \n",
       "2  REP.A001_A549_24H_X1_B27|REP.A001_A549_24H_X2_...       A03     0   \n",
       "3  REP.A001_A549_24H_X1_B27|REP.A001_A549_24H_X2_...       A04     0   \n",
       "4  REP.A001_A549_24H_X1_B27|REP.A001_A549_24H_X2_...       A04     0   \n",
       "\n",
       "   Metadata_broad_sample  pert_iname              moa  \n",
       "0                   DMSO        DMSO  Control vehicle  \n",
       "1                   DMSO        DMSO  Control vehicle  \n",
       "2                   DMSO        DMSO  Control vehicle  \n",
       "3                   DMSO        DMSO  Control vehicle  \n",
       "4                   DMSO        DMSO  Control vehicle  \n",
       "\n",
       "[5 rows x 988 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_level4.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(27837, 988)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_level4.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "128"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df_level4['det_plate'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "380"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df_level4['det_well'].unique())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### - Remove highly correlated landmark genes and samples with Null compound values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def feature_selection(df_data):\n",
    "    \n",
    "    \"\"\"\n",
    "    Perform feature selection by dropping columns with null compounds values, \n",
    "    and highly correlated landmark genes from the data.\n",
    "    \"\"\"\n",
    "    \n",
    "    df_data_genes = df_data.drop(['replicate_id', 'Metadata_broad_sample', 'pert_id', 'dose', 'pert_idose', \n",
    "                                  'pert_iname', 'moa', 'sig_id', 'det_plate', 'det_well'], axis = 1).copy()\n",
    "    df_data_corr = df_data_genes.corr(method = 'spearman')\n",
    "    drop_cols = []\n",
    "    n_cols = len(df_data_corr.columns)\n",
    "    for i in range(n_cols):\n",
    "        for k in range(i+1, n_cols):\n",
    "            val = df_data_corr.iloc[k, i]\n",
    "            col = df_data_corr.columns[i]\n",
    "            if abs(val) >= 0.9:\n",
    "                drop_cols.append(col)\n",
    "    df_data.drop(set(drop_cols), axis = 1, inplace = True)\n",
    "    df_data.drop(df_data[df_data['pert_iname'].isnull()].index).reset_index(drop = True, inplace = True)\n",
    "    \n",
    "    return df_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_level4 = feature_selection(df_level4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(27837, 988)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_level4.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_median_score(cpds_list, df):\n",
    "    \"\"\"\n",
    "    This function calculates the median score for each compound based on its replicates\n",
    "    \"\"\"\n",
    "    \n",
    "    cpds_median_score = {}\n",
    "    for cpd in cpds_list:\n",
    "        cpd_replicates = df[df['pert_iname'] == cpd].copy()\n",
    "        cpd_replicates.drop(['replicate_id', 'Metadata_broad_sample', 'pert_id', 'dose', 'pert_idose', \n",
    "                             'pert_iname', 'moa', 'sig_id', 'det_plate', 'det_well'], axis = 1, inplace = True)\n",
    "        cpd_replicates_corr = cpd_replicates.astype('float64').T.corr(method = 'spearman').values\n",
    "        if len(cpd_replicates_corr) == 1:\n",
    "            median_val = 1\n",
    "        else:\n",
    "            median_val = median(list(cpd_replicates_corr[np.triu_indices(len(cpd_replicates_corr), k = 1)]))\n",
    "        \n",
    "        cpds_median_score[cpd] = median_val\n",
    "        \n",
    "    return cpds_median_score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_compounds(cpd_med_score, df):\n",
    "    \"\"\"\n",
    "    Check if all distinct compounds in the Level-4 dataframe are present \n",
    "    in the cpd_med_score dictionary, if not add the compounds as keys to the dictionary \n",
    "    and give them a null value.\n",
    "    \"\"\"\n",
    "    cpd_list = df['pert_iname'].unique().tolist()\n",
    "    cpd_keys = cpd_med_score.keys()\n",
    "    for cpd in cpd_list:\n",
    "        if cpd not in cpd_keys:\n",
    "            cpd_med_score[cpd] = np.nan\n",
    "            \n",
    "    return cpd_med_score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_cpd_medianscores(df):\n",
    "    \n",
    "    \"\"\"This function computes median scores for all compounds found in the Level-4 dataframe PER DOSE (1-6)\"\"\"\n",
    "    \n",
    "    dose_list = list(set(df['dose'].unique().tolist()))[1:7]\n",
    "    \n",
    "    for dose in dose_list:\n",
    "        df_dose = df[df['dose'] == dose].copy()\n",
    "        cpds_list = df_dose['pert_iname'].unique().tolist()\n",
    "        cpds_median_score = get_median_score(cpds_list, df_dose)\n",
    "        cpds_median_score = check_compounds(cpds_median_score, df)\n",
    "        sorted_med_score = {key:value for key, value in sorted(cpds_median_score.items(), key=lambda item: item[0])}\n",
    "        if dose == 1:\n",
    "            df_cpd_med_score = pd.DataFrame.from_dict(sorted_med_score, orient='index', columns = ['dose_1'])\n",
    "        else:\n",
    "            df_cpd_med_score['dose_' + str(dose)] = sorted_med_score.values()\n",
    "            \n",
    "    return df_cpd_med_score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_cpd_med_score = get_cpd_medianscores(df_level4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dose_1</th>\n",
       "      <th>dose_2</th>\n",
       "      <th>dose_3</th>\n",
       "      <th>dose_4</th>\n",
       "      <th>dose_5</th>\n",
       "      <th>dose_6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>16,16-dimethylprostaglandin-e2</th>\n",
       "      <td>0.174516</td>\n",
       "      <td>0.078331</td>\n",
       "      <td>0.299300</td>\n",
       "      <td>0.081933</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17-hydroxyprogesterone-caproate</th>\n",
       "      <td>0.024738</td>\n",
       "      <td>0.037952</td>\n",
       "      <td>0.067445</td>\n",
       "      <td>0.145491</td>\n",
       "      <td>0.096967</td>\n",
       "      <td>0.180514</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2-iminobiotin</th>\n",
       "      <td>-0.133693</td>\n",
       "      <td>0.237833</td>\n",
       "      <td>-0.043386</td>\n",
       "      <td>-0.032912</td>\n",
       "      <td>0.099408</td>\n",
       "      <td>0.099065</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3,3'-diindolylmethane</th>\n",
       "      <td>0.036497</td>\n",
       "      <td>0.014831</td>\n",
       "      <td>-0.007878</td>\n",
       "      <td>-0.016737</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3-amino-benzamide</th>\n",
       "      <td>0.051959</td>\n",
       "      <td>-0.015037</td>\n",
       "      <td>0.001821</td>\n",
       "      <td>-0.010001</td>\n",
       "      <td>-0.011279</td>\n",
       "      <td>0.111988</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3-deazaadenosine</th>\n",
       "      <td>0.065967</td>\n",
       "      <td>-0.059889</td>\n",
       "      <td>0.048698</td>\n",
       "      <td>-0.011885</td>\n",
       "      <td>-0.024411</td>\n",
       "      <td>0.090736</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABT-737</th>\n",
       "      <td>0.016400</td>\n",
       "      <td>0.017630</td>\n",
       "      <td>0.036690</td>\n",
       "      <td>0.020696</td>\n",
       "      <td>0.077362</td>\n",
       "      <td>0.151709</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AICA-ribonucleotide</th>\n",
       "      <td>-0.009703</td>\n",
       "      <td>0.091991</td>\n",
       "      <td>0.033783</td>\n",
       "      <td>0.102765</td>\n",
       "      <td>0.010875</td>\n",
       "      <td>0.055641</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AKT-inhibitor-1-2</th>\n",
       "      <td>-0.008655</td>\n",
       "      <td>0.139537</td>\n",
       "      <td>0.090489</td>\n",
       "      <td>0.075586</td>\n",
       "      <td>-0.033984</td>\n",
       "      <td>0.043067</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ALX-5407</th>\n",
       "      <td>0.068434</td>\n",
       "      <td>0.062257</td>\n",
       "      <td>0.093194</td>\n",
       "      <td>0.007703</td>\n",
       "      <td>0.000444</td>\n",
       "      <td>0.072881</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   dose_1    dose_2    dose_3    dose_4  \\\n",
       "16,16-dimethylprostaglandin-e2   0.174516  0.078331  0.299300  0.081933   \n",
       "17-hydroxyprogesterone-caproate  0.024738  0.037952  0.067445  0.145491   \n",
       "2-iminobiotin                   -0.133693  0.237833 -0.043386 -0.032912   \n",
       "3,3'-diindolylmethane            0.036497  0.014831 -0.007878 -0.016737   \n",
       "3-amino-benzamide                0.051959 -0.015037  0.001821 -0.010001   \n",
       "3-deazaadenosine                 0.065967 -0.059889  0.048698 -0.011885   \n",
       "ABT-737                          0.016400  0.017630  0.036690  0.020696   \n",
       "AICA-ribonucleotide             -0.009703  0.091991  0.033783  0.102765   \n",
       "AKT-inhibitor-1-2               -0.008655  0.139537  0.090489  0.075586   \n",
       "ALX-5407                         0.068434  0.062257  0.093194  0.007703   \n",
       "\n",
       "                                   dose_5    dose_6  \n",
       "16,16-dimethylprostaglandin-e2        NaN       NaN  \n",
       "17-hydroxyprogesterone-caproate  0.096967  0.180514  \n",
       "2-iminobiotin                    0.099408  0.099065  \n",
       "3,3'-diindolylmethane                 NaN       NaN  \n",
       "3-amino-benzamide               -0.011279  0.111988  \n",
       "3-deazaadenosine                -0.024411  0.090736  \n",
       "ABT-737                          0.077362  0.151709  \n",
       "AICA-ribonucleotide              0.010875  0.055641  \n",
       "AKT-inhibitor-1-2               -0.033984  0.043067  \n",
       "ALX-5407                         0.000444  0.072881  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_cpd_med_score.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "def drop_cpds_with_null(df):\n",
    "    \"\"\"\n",
    "    This function drop compounds with median scores of 1 \n",
    "    or null values in any of the dose points (1-6)\n",
    "    \"\"\"\n",
    "    cpds_with_null = []\n",
    "    for cpd in df.index:\n",
    "        if any(df.loc[cpd] == 1) | any(df.loc[cpd].isnull()):\n",
    "            cpds_with_null.append(cpd)\n",
    "    df.drop(cpds_with_null, axis = 0, inplace = True)\n",
    "    \n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_cpd_med_score = drop_cpds_with_null(df_cpd_med_score)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1330, 6)"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_cpd_med_score.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dose_1</th>\n",
       "      <th>dose_2</th>\n",
       "      <th>dose_3</th>\n",
       "      <th>dose_4</th>\n",
       "      <th>dose_5</th>\n",
       "      <th>dose_6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>17-hydroxyprogesterone-caproate</th>\n",
       "      <td>0.024738</td>\n",
       "      <td>0.037952</td>\n",
       "      <td>0.067445</td>\n",
       "      <td>0.145491</td>\n",
       "      <td>0.096967</td>\n",
       "      <td>0.180514</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2-iminobiotin</th>\n",
       "      <td>-0.133693</td>\n",
       "      <td>0.237833</td>\n",
       "      <td>-0.043386</td>\n",
       "      <td>-0.032912</td>\n",
       "      <td>0.099408</td>\n",
       "      <td>0.099065</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3-amino-benzamide</th>\n",
       "      <td>0.051959</td>\n",
       "      <td>-0.015037</td>\n",
       "      <td>0.001821</td>\n",
       "      <td>-0.010001</td>\n",
       "      <td>-0.011279</td>\n",
       "      <td>0.111988</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3-deazaadenosine</th>\n",
       "      <td>0.065967</td>\n",
       "      <td>-0.059889</td>\n",
       "      <td>0.048698</td>\n",
       "      <td>-0.011885</td>\n",
       "      <td>-0.024411</td>\n",
       "      <td>0.090736</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABT-737</th>\n",
       "      <td>0.016400</td>\n",
       "      <td>0.017630</td>\n",
       "      <td>0.036690</td>\n",
       "      <td>0.020696</td>\n",
       "      <td>0.077362</td>\n",
       "      <td>0.151709</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AICA-ribonucleotide</th>\n",
       "      <td>-0.009703</td>\n",
       "      <td>0.091991</td>\n",
       "      <td>0.033783</td>\n",
       "      <td>0.102765</td>\n",
       "      <td>0.010875</td>\n",
       "      <td>0.055641</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AKT-inhibitor-1-2</th>\n",
       "      <td>-0.008655</td>\n",
       "      <td>0.139537</td>\n",
       "      <td>0.090489</td>\n",
       "      <td>0.075586</td>\n",
       "      <td>-0.033984</td>\n",
       "      <td>0.043067</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ALX-5407</th>\n",
       "      <td>0.068434</td>\n",
       "      <td>0.062257</td>\n",
       "      <td>0.093194</td>\n",
       "      <td>0.007703</td>\n",
       "      <td>0.000444</td>\n",
       "      <td>0.072881</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AS-605240</th>\n",
       "      <td>-0.206672</td>\n",
       "      <td>0.071962</td>\n",
       "      <td>-0.179716</td>\n",
       "      <td>-0.187534</td>\n",
       "      <td>0.175653</td>\n",
       "      <td>0.093238</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AT-7519</th>\n",
       "      <td>0.123918</td>\n",
       "      <td>-0.010591</td>\n",
       "      <td>0.073419</td>\n",
       "      <td>0.441535</td>\n",
       "      <td>0.822246</td>\n",
       "      <td>0.850807</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   dose_1    dose_2    dose_3    dose_4  \\\n",
       "17-hydroxyprogesterone-caproate  0.024738  0.037952  0.067445  0.145491   \n",
       "2-iminobiotin                   -0.133693  0.237833 -0.043386 -0.032912   \n",
       "3-amino-benzamide                0.051959 -0.015037  0.001821 -0.010001   \n",
       "3-deazaadenosine                 0.065967 -0.059889  0.048698 -0.011885   \n",
       "ABT-737                          0.016400  0.017630  0.036690  0.020696   \n",
       "AICA-ribonucleotide             -0.009703  0.091991  0.033783  0.102765   \n",
       "AKT-inhibitor-1-2               -0.008655  0.139537  0.090489  0.075586   \n",
       "ALX-5407                         0.068434  0.062257  0.093194  0.007703   \n",
       "AS-605240                       -0.206672  0.071962 -0.179716 -0.187534   \n",
       "AT-7519                          0.123918 -0.010591  0.073419  0.441535   \n",
       "\n",
       "                                   dose_5    dose_6  \n",
       "17-hydroxyprogesterone-caproate  0.096967  0.180514  \n",
       "2-iminobiotin                    0.099408  0.099065  \n",
       "3-amino-benzamide               -0.011279  0.111988  \n",
       "3-deazaadenosine                -0.024411  0.090736  \n",
       "ABT-737                          0.077362  0.151709  \n",
       "AICA-ribonucleotide              0.010875  0.055641  \n",
       "AKT-inhibitor-1-2               -0.033984  0.043067  \n",
       "ALX-5407                         0.000444  0.072881  \n",
       "AS-605240                        0.175653  0.093238  \n",
       "AT-7519                          0.822246  0.850807  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_cpd_med_score.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "def no_of_replicates_per_cpd(df, df_lvl4):\n",
    "    \"\"\"This function computes the numbers of replicates for each compound (cpd_size)\"\"\"\n",
    "    \n",
    "    dose_list = list(set(df_lvl4['dose'].unique().tolist()))[1:7]\n",
    "    cpds_no_of_reps = {}\n",
    "    for cpd in df.index:\n",
    "        num_of_reps = 0\n",
    "        df_cpd = df_lvl4[df_lvl4['pert_iname'] == cpd].copy()\n",
    "        for dose in dose_list:\n",
    "            df_dose = df_cpd[df_cpd['dose'] == dose].copy()\n",
    "            num_of_reps += df_dose.shape[0]\n",
    "        cpds_no_of_reps[cpd] = num_of_reps // len(dose_list)\n",
    "    df['no_of_replicates'] = cpds_no_of_reps.values()\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_cpd_med_score = no_of_replicates_per_cpd(df_cpd_med_score, df_level4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dose_1</th>\n",
       "      <th>dose_2</th>\n",
       "      <th>dose_3</th>\n",
       "      <th>dose_4</th>\n",
       "      <th>dose_5</th>\n",
       "      <th>dose_6</th>\n",
       "      <th>no_of_replicates</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>17-hydroxyprogesterone-caproate</th>\n",
       "      <td>0.024738</td>\n",
       "      <td>0.037952</td>\n",
       "      <td>0.067445</td>\n",
       "      <td>0.145491</td>\n",
       "      <td>0.096967</td>\n",
       "      <td>0.180514</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2-iminobiotin</th>\n",
       "      <td>-0.133693</td>\n",
       "      <td>0.237833</td>\n",
       "      <td>-0.043386</td>\n",
       "      <td>-0.032912</td>\n",
       "      <td>0.099408</td>\n",
       "      <td>0.099065</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3-amino-benzamide</th>\n",
       "      <td>0.051959</td>\n",
       "      <td>-0.015037</td>\n",
       "      <td>0.001821</td>\n",
       "      <td>-0.010001</td>\n",
       "      <td>-0.011279</td>\n",
       "      <td>0.111988</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3-deazaadenosine</th>\n",
       "      <td>0.065967</td>\n",
       "      <td>-0.059889</td>\n",
       "      <td>0.048698</td>\n",
       "      <td>-0.011885</td>\n",
       "      <td>-0.024411</td>\n",
       "      <td>0.090736</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABT-737</th>\n",
       "      <td>0.016400</td>\n",
       "      <td>0.017630</td>\n",
       "      <td>0.036690</td>\n",
       "      <td>0.020696</td>\n",
       "      <td>0.077362</td>\n",
       "      <td>0.151709</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AICA-ribonucleotide</th>\n",
       "      <td>-0.009703</td>\n",
       "      <td>0.091991</td>\n",
       "      <td>0.033783</td>\n",
       "      <td>0.102765</td>\n",
       "      <td>0.010875</td>\n",
       "      <td>0.055641</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AKT-inhibitor-1-2</th>\n",
       "      <td>-0.008655</td>\n",
       "      <td>0.139537</td>\n",
       "      <td>0.090489</td>\n",
       "      <td>0.075586</td>\n",
       "      <td>-0.033984</td>\n",
       "      <td>0.043067</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ALX-5407</th>\n",
       "      <td>0.068434</td>\n",
       "      <td>0.062257</td>\n",
       "      <td>0.093194</td>\n",
       "      <td>0.007703</td>\n",
       "      <td>0.000444</td>\n",
       "      <td>0.072881</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AS-605240</th>\n",
       "      <td>-0.206672</td>\n",
       "      <td>0.071962</td>\n",
       "      <td>-0.179716</td>\n",
       "      <td>-0.187534</td>\n",
       "      <td>0.175653</td>\n",
       "      <td>0.093238</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AT-7519</th>\n",
       "      <td>0.123918</td>\n",
       "      <td>-0.010591</td>\n",
       "      <td>0.073419</td>\n",
       "      <td>0.441535</td>\n",
       "      <td>0.822246</td>\n",
       "      <td>0.850807</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   dose_1    dose_2    dose_3    dose_4  \\\n",
       "17-hydroxyprogesterone-caproate  0.024738  0.037952  0.067445  0.145491   \n",
       "2-iminobiotin                   -0.133693  0.237833 -0.043386 -0.032912   \n",
       "3-amino-benzamide                0.051959 -0.015037  0.001821 -0.010001   \n",
       "3-deazaadenosine                 0.065967 -0.059889  0.048698 -0.011885   \n",
       "ABT-737                          0.016400  0.017630  0.036690  0.020696   \n",
       "AICA-ribonucleotide             -0.009703  0.091991  0.033783  0.102765   \n",
       "AKT-inhibitor-1-2               -0.008655  0.139537  0.090489  0.075586   \n",
       "ALX-5407                         0.068434  0.062257  0.093194  0.007703   \n",
       "AS-605240                       -0.206672  0.071962 -0.179716 -0.187534   \n",
       "AT-7519                          0.123918 -0.010591  0.073419  0.441535   \n",
       "\n",
       "                                   dose_5    dose_6  no_of_replicates  \n",
       "17-hydroxyprogesterone-caproate  0.096967  0.180514                 3  \n",
       "2-iminobiotin                    0.099408  0.099065                 2  \n",
       "3-amino-benzamide               -0.011279  0.111988                 3  \n",
       "3-deazaadenosine                -0.024411  0.090736                 2  \n",
       "ABT-737                          0.077362  0.151709                 3  \n",
       "AICA-ribonucleotide              0.010875  0.055641                 3  \n",
       "AKT-inhibitor-1-2               -0.033984  0.043067                 3  \n",
       "ALX-5407                         0.000444  0.072881                 3  \n",
       "AS-605240                        0.175653  0.093238                 2  \n",
       "AT-7519                          0.822246  0.850807                 3  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_cpd_med_score.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "def save_to_csv(df, path, file_name, compress=None):\n",
    "    \"\"\"saves dataframes to csv\"\"\"\n",
    "    \n",
    "    if not os.path.exists(path):\n",
    "        os.mkdir(path)\n",
    "    \n",
    "    df.to_csv(os.path.join(path, file_name), index=False, compression=compress)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "save_to_csv(df_cpd_med_score.reset_index().rename({'index':'cpd'}, axis = 1), \n",
    "            'L1000_lvl4_cpd_replicate_datasets', 'cpd_replicate_median_scores.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "save_to_csv(df_level4, 'L1000_lvl4_cpd_replicate_datasets', \n",
    "            'L1000_level4_cpd_replicates.csv.gz', compress=\"gzip\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
